starcheck/calc_ccd_temps.py

```diff
@@ -150,41 +164,65 @@ def get_ccd_temps(oflsdir, outdir='out',
-    proc['datestop'] = DateTime(tstop).date
-
-    # Get temperature telemetry for 1 days prior to
-    # min(last available telem, backstop tstart, run_start_time)
-    # where run_start_time is for regression testing.
-    end_time = fetch.get_time_range('aacccdpt', format='secs')[1]
-    tlm = get_telem_values(min(end_time, tstart, run_start_time.secs),
-                           ['aacccdpt'],
-                           days=1)
-
-    states = get_week_states(tstart, tstop, bs_cmds, tlm)
-
-    # If the last obsid interval extends over the end of states
-    # extend the state / predictions
-    if ((states[-1]['obsid'] == sc_obsids[-1]['obsid']) &
-            (sc_obsids[-1]['obs_tstop'] > states[-1]['tstop'])):
-        tstop = sc_obsids[-1]['obs_tstop']
-        states[-1]['tstop'] = sc_obsids[-1]['obs_tstop']
-        states[-1]['datestop'] = DateTime(sc_obsids[-1]['obs_tstop']).date
-
-    if tstart > DateTime(MODEL_VALID_FROM).secs:
-        times, ccd_temp = make_week_predict(model_spec, states, tstop)
+    bs_dates = bs_cmds['date']
+
+    # Running loads termination time is the last time of "current running loads"
+    # (or in the case of a safing action, "current approved load commands" in
+    # kadi commands) which should be included in propagation. Starting from
+    # around 2020-April this is included as a commmand in the loads, while prior
+    # to that we just use the first command in the backstop loads.
+    ok = bs_cmds['event_type'] == 'RUNNING_LOAD_TERMINATION_TIME'
+    rltt = DateTime(bs_dates[ok][0] if np.any(ok) else bs_dates[0])
+
+    # First actual command in backstop loads (all the NOT-RLTT commands)
+    bs_start = DateTime(bs_dates[~ok][0])
+
+    # Scheduled stop time is the end of propagation, either the explicit
+    # time as a pseudo-command in the loads or the last backstop command time.
+    ok = bs_cmds['event_type'] == 'SCHEDULED_STOP_TIME'
+    sched_stop = DateTime(bs_dates[ok][0] if np.any(ok) else bs_dates[-1])
+
+    if 'test_rltt' in kwargs:
+        rltt = DateTime(kwargs['test_rltt'])
+    if 'test_sched_stop' in kwargs:
+        sched_stop = DateTime(kwargs['test_sched_stop'])
+
+    logger.info(f'RLTT = {rltt.date}')
+    logger.info(f'sched_stop = {sched_stop.date}')
+
+    proc['datestart'] = bs_start.date
+    proc['datestop'] = sched_stop.date
+
+    # Get temperature telemetry for 1 day prior to min(last available telem,
+    # backstop start, run_start_time) where run_start_time is for regression
+    # testing.
+    tlm_end_time = min(fetch.get_time_range('aacccdpt', format='secs')[1],
+                       bs_start.secs, run_start_time.secs)
+    tlm = get_telem_values(tlm_end_time, ['aacccdpt'], days=1)
+    states = get_week_states(rltt, sched_stop, bs_cmds, tlm)
+
+    # If the last obsid interval extends over the end of states then
+    # extend the state / predictions.
+    last_state = states[-1]
+    last_sc_obsid = sc_obsids[-1]
```

Comment:
I don't understand the circumstances where starcheck would make an obsid interval that extends beyond the end of commanding.  With SCHEDULED_STOP in backstop this bit is not needed, but wondering if it is needed for regression testing or whatnot.

Reply:
IIRC in the absence of something useful like SCHEDULED_STOP, if the schedule ended in NPNT (and had no maneuver in backstop to define end time), the obsid stop time for the last observation in the schedule was sometimes set from the stop time listed in the processing summary?  Maybe? 

I remember spending a lot of time on corner cases to get it to not just fail out.  Should be moot if we can do a modern hopper transition and better state building but there you go.

Reply:
I added your text as a comment to the code for our future selves.